lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,969 @@
+  void Post(const parser::CaseConstruct &caseConstruct) {
+    PopName(caseConstruct);
+  }
+  void Post(const parser::SelectRankConstruct &selectRankConstruct) {
+    PopName(selectRankConstruct);
+  }
+  void Post(const parser::SelectTypeConstruct &selectTypeConstruct) {
+    PopName(selectTypeConstruct);
+  }
+
+  //  Named constructs without block scope
+  void Post(const parser::WhereConstruct &whereConstruct) {
+    PopNonBlockConstructName(whereConstruct);
+  }
+  void Post(const parser::ForallConstruct &forallConstruct) {
+    PopNonBlockConstructName(forallConstruct);
+  }
+
+  //  Statements with label references
+  void Post(const parser::LabelDoStmt &labelDoStmt) {
+    AddDoBase(std::get<parser::Label>(labelDoStmt.t));
+  }
+  void Post(const parser::GotoStmt &gotoStmt) { AddBase(gotoStmt.v); }
+  void Post(const parser::ComputedGotoStmt &computedGotoStmt) {
+    AddBase(std::get<std::list<parser::Label>>(computedGotoStmt.t));
+  }
+  void Post(const parser::ArithmeticIfStmt &arithmeticIfStmt) {
+    AddBase(std::get<1>(arithmeticIfStmt.t));
+    AddBase(std::get<2>(arithmeticIfStmt.t));
+    AddBase(std::get<3>(arithmeticIfStmt.t));
+  }
+  void Post(const parser::AssignStmt &assignStmt) {
+    AddBase(std::get<parser::Label>(assignStmt.t));
+  }
+  void Post(const parser::AssignedGotoStmt &assignedGotoStmt) {
+    AddBase(std::get<std::list<parser::Label>>(assignedGotoStmt.t));
+  }
+  void Post(const parser::AltReturnSpec &altReturnSpec) {
+    AddBase(altReturnSpec.v);
+  }
+
+  void Post(const parser::ErrLabel &errLabel) { AddBase(errLabel.v); }
+  void Post(const parser::EndLabel &endLabel) { AddBase(endLabel.v); }
+  void Post(const parser::EorLabel &eorLabel) { AddBase(eorLabel.v); }
+  void Post(const parser::Format &format) {
+    // BUG: the label is saved as an IntLiteralConstant rather than a Label
+#if 0
+    if (const auto *P{std::get_if<parser::Label>(&format.u)}) {
+      AddFmtBase(*P);
+    }
+#else
+    // FIXME: this is wrong, but extracts the label's value
+    if (const auto *P{std::get_if<0>(&format.u)}) {
+      AddFmtBase(parser::Label{std::get<0>(std::get<parser::IntLiteralConstant>(
+          std::get<parser::LiteralConstant>((*P->thing).u).u)
+                                               .t)});
+    }
+#endif
+  }
+  void Post(const parser::CycleStmt &cycleStmt) {
+    if (cycleStmt.v.has_value()) {
+      CheckLabelContext("CYCLE", cycleStmt.v.value().ToString());
+    }
+  }
+  void Post(const parser::ExitStmt &exitStmt) {
+    if (exitStmt.v.has_value()) {
+      CheckLabelContext("EXIT", exitStmt.v.value().ToString());
+    }
+  }
+
+  // Getters for the results
+  const std::vector<UnitAnalysis> &GetProgramUnits() const {
+    return programUnits_;
+  }
+  ErrorHandler &GetEH() { return eh; }
+
+private:
+  bool PushScope() {
+    programUnits_.back().scopeModel_.push_back(currentScope_);
+    currentScope_ = programUnits_.back().scopeModel_.size() - 1;
```

Comment:
Maybe `scopeModel_` could be a `std::stack` and then you could just mess with its top element rather than a `currentScope_` index?

Reply:
I was at a conference about 30 years ago (OOPSLA?) when Alex Stepanov gave a keynote. In that keynote, he (paraphrasing) said, he regretted having been coerced into adding std::stack to the STL. In summary, "no one should ever use std::stack".